sklearn/utils/_pprint.py

```diff
@@ -96,8 +96,12 @@ def _changed_params(estimator):
     init_params = {name: param.default for name, param in init_params.items()}
 
     for k, v in params.items():
-        if (k not in init_params or (  # happens if k is part of a **kwargs
-                repr(v) != repr(init_params[k]) and
+        if (k not in init_params or  # happens if k is part of a **kwargs
+            init_params[k] == _empty or  # k has no default value
+            # don't `repr` nested estimators, it causes nonlinear complexity
+            (isinstance(v, BaseEstimator) and
+                v.__class__ != init_params[k].__class__) or
+            (repr(v) != repr(init_params[k]) and
```

Comment:
Maybe we should just declare a small helper just above, to help with readability:

```py

def has_changed(k, v):
	if k not in init_params:
		# happens if k is part of a **kwargs
		return True
	if ...	
```

Reply:
Yes, printing time is now roughly the same and `__repr__` of `BaseEstimator` is called the same amount of times whether `print_changed_only` is `True` or `False`.

Although, I don't understand why the `mypy` check fails with error `sklearn/utils/_pprint.py:66: error: Module 'inspect' has no attribute '_empty'`.

I think the helper function is a good idea, I will implement it.

Reply:
maybe just import `inspect` and use `inspect.signature` and `inspect._empty`

If that doesn't work, try adding a `# type: ignore` comment on the line.